File: sdk/ts/core/signing-transport.ts
// sdk/ts outbound auto-sign transport — the TS sibling of Go
// core.NewSigningTransport (sdk/go/core/transport.go) and Python
// SigningTransport / SignedOutbound (sdk/python/ramp_sdk/signing_transport.py).
//
// Core Invariant: this module is a pure ORCHESTRATION of the already
// byte-parity-locked primitives signRequest / appendSignature (core/sign-request.ts)
// and clockWindow / monotonicWindow (core/window.ts). It stamps EVERY covered header
// at the value that entered the signature base — Content-Digest / Signature-Input /
// Signature, and the Authorization / Signature-Agent whose values may be empty —
// byte-identical to the shared Go/Python oracle, forwards the request body UNMODIFIED,
// and adds NO new crypto and NO new signature-base rendering.
//
// Two faces, mirroring Python's transport-neutral shape (the cleaner fit for TS's
// multi-runtime edge, which has no Go http.RoundTripper):
//   - signOutbound(...): the transport-NEUTRAL header core (Python SignedOutbound
//     sibling) — computes the RFC 9421 headers for one request and returns them
//     with the untouched body. Wraps no client.
//   - createSigningTransport(send, opts): wraps a WHATWG-fetch-shaped outbound seam
//     send(url, init) — buffers the body, computes headers via signOutbound, and
//     forwards the SAME body bytes to send.
//
// Signature-Agent is SET-IF-ABSENT (mirror Go transport.go:143-145, NOT Python's
// always-stamp): TS carries the relay/append chain, where an upstream sig1 already
// covers its own Signature-Agent value and must never be overwritten.

import { SignatureAgentHeader } from "../src/wire.ts";
import {
	appendSignature,
	type PriorSignatures,
	type SignRequestOptions,
	signRequest,
} from "./sign-request.ts";
import { clockWindow, type Window } from "./window.ts";

// The DEFAULT freshness window: clock-derived created + a 5-minute TTL, matching
// Go's `signWindow = 5 * time.Minute` (transport.go:18) — NOT Python's 600s. The
// clock reads seconds (Date.now()/1000); clockWindow floors to Go's .Unix().
const DEFAULT_WINDOW_TTL_SEC = 300;

function defaultWindow(): Window {
	return clockWindow(() => Date.now() / 1000, DEFAULT_WINDOW_TTL_SEC);
}

// Case-insensitive header lookup over a plain header record. Incoming requests may
// spell header names in any case; the covered values (authorization, prior
// Signature state, Signature-Agent) must be read regardless of casing.
function getHeader(
	headers: Record<string, string>,
	name: string,
): string | undefined {
	const lower = name.toLowerCase();
	for (const k of Object.keys(headers)) {
		if (k.toLowerCase() === lower) return headers[k];
	}
	return undefined;
}

/**
 * mergeSigned overlays the signed headers onto the caller's, REPLACING any the caller
 * spelled in a different case rather than letting both survive.
 *
 * A plain spread would not: header names are case-insensitive on the wire, JS object keys
 * are not, so a caller's `Authorization` and a signed `authorization` become two field
 * lines. getHeader above already reads case-insensitively; this is the write side
 * agreeing with it. See docs/design-history.md, "A covered header the peer never receives
 * is not bound".
 */
function mergeSigned(
	callerHeaders: Record<string, string>,
	signedHeaders: Record<string, string>,
): Record<string, string> {
	const claimed = new Set(Object.keys(signedHeaders).map((k) => k.toLowerCase()));
	const merged: Record<string, string> = {};
	for (const [name, value] of Object.entries(callerHeaders)) {
		if (!claimed.has(name.toLowerCase())) merged[name] = value;
	}
	return { ...merged, ...signedHeaders };
}

/** Inputs for signOutbound — the transport-neutral header core. */
export interface SignOutboundOptions {
	privKey: CryptoKey;
	keyid: string;
	method: string;
	url: string;
	// Uint8Array<ArrayBuffer> (never SharedArrayBuffer-backed), matching the
	// sdk/ts WebCrypto convention (core/sign-request.ts, core/verifier.ts).
	body: Uint8Array<ArrayBuffer>;
	authorization: string;
	// The COVERED Signature-Agent value to bind ("" binds an empty header, the
	// no-directory bootstrap path). The transport resolves set-if-absent BEFORE
	// calling — this is the already-resolved value.
	signatureAgent: string;
	// Freshness window; defaults to clockWindow(now, 300s) (Go 5m) when absent.
	window?: Window;
	// appendOnly routes through appendSignature even for a fresh request (relay
	// mode); AppendSignature degrades to a byte-identical sig1 when prior is empty.
	appendOnly?: boolean;
	// Prior signature state carried by the incoming request (the relay/chain path).
	prior?: PriorSignatures;
}

/** The RFC 9421 header set (plus the untouched body) a signed request carries. */
export interface SignedOutbound {
	headers: Record<string, string>;
	body: Uint8Array<ArrayBuffer>;
}

/**
 * signOutbound computes the RFC 9421 Content-Digest / Signature-Input / Signature
 * headers for one outbound request and returns them with the body untouched — the
 * transport-neutral core (Python SignedOutbound sibling). It orchestrates the
 * parity-locked primitives: appendSignature when appendOnly is set OR a prior
 * Signature is present (forwarding chain), signRequest otherwise.
 *
 * It returns EVERY covered header at the value that entered the signature base, empty
 * values included — see the emit below for why. The keys are LOWERCASE, which is the
 * spelling the corpus records and Python emits, and is what lets a merge over a caller's
 * own headers replace rather than duplicate.
 */
export async function signOutbound(
	o: SignOutboundOptions,
): Promise<SignedOutbound> {
	const [created, expires] = (o.window ?? defaultWindow())();
	const signOpts: SignRequestOptions = {
		method: o.method,
		url: o.url,
		body: o.body,
		authorization: o.authorization,
		signatureAgent: o.signatureAgent,
		keyid: o.keyid,
		created,
		expires,
	};
	const prior = o.prior ?? { signatureInput: "", signature: "" };
	const chained = (o.appendOnly ?? false) || prior.signature !== "";
	const signed = chained
		? await appendSignature(o.privKey, prior, signOpts)
		: await signRequest(o.privKey, signOpts);

	// EVERY covered header is emitted, at exactly the value that entered the signature
	// base — empty values included. A verifier rebuilds the base from the request it
	// received, so a value bound but never sent is not bound at all: it reads the covered
	// names off signature-input, finds nothing on the wire under one of them, and refuses.
	// Measured: the covered set binds authorization and signature-agent unconditionally,
	// and a request carrying neither is answered `header "authorization" missing from
	// request` by every conformant verifier — so the ports agreed byte-for-byte with the
	// oracle on what they signed and could not complete a single call.
	//
	// EVERY covered header, at exactly the value that entered the signature base — empty
	// values included. See docs/design-history.md, "A covered header the peer never
	// receives is not bound", for why binding one without sending it is not binding it,
	// and why the emitted key is LOWERCASE (a signed key spelled differently from the
	// caller's survives the merge beside it, putting the name on the wire twice).
	//
	// Taken straight off `signed`, never re-read from `o`: the primitive echoes what it
	// bound, so there is one place the emitted value can come from and no way for the two
	// to drift.
	const headers: Record<string, string> = {
		"content-digest": signed.contentDigest,
		"signature-input": signed.signatureInput,
		signature: signed.signature,
		authorization: signed.authorization,
		[SignatureAgentHeader.toLowerCase()]: signed.signatureAgent,
	};
	return { headers, body: o.body };
}

/** The WHATWG-fetch-shaped outbound request the transport inspects / forwards. */
export interface OutboundInit {
	method?: string;
	headers?: Record<string, string>;
	body?: Uint8Array<ArrayBuffer>;
}

/** The seam createSigningTransport wraps: a WHATWG-fetch-shaped send(url, init). */
export type OutboundSend<R> = (url: string, init: OutboundInit) => Promise<R>;

/** What a sign predicate inspects to decide whether a request is signed. */
export interface OutboundRequest {
	url: string;
	method: string;
	headers: Record<string, string>;
	body: Uint8Array<ArrayBuffer> | undefined;
}

/**
 * Options for createSigningTransport — an idiomatic TS options object, one field per
 * Go WithX (transport.go:47-96). window replaces the default freshness window;
 * appendOnly forces the relay append branch; signatureAgent supplies the covered
 * directory value stamped SET-IF-ABSENT; predicate gates which requests are signed
 * (default: sign every bodied request).
 */
export interface SigningTransportOptions {
	privKey: CryptoKey;
	keyid: string;
	window?: Window;
	appendOnly?: boolean;
	signatureAgent?: string;
	predicate?: (req: OutboundRequest) => boolean;
}

/**
 * createSigningTransport wraps a WHATWG-fetch-shaped send and returns a send with the
 * same shape that auto-signs each outbound request: it buffers the body, computes
 * the RFC 9421 headers via signOutbound, merges them, and forwards the SAME body
 * bytes to the wrapped send. A request with no body — or one the predicate
 * excludes — passes through UNSIGNED (there is nothing to bind a Content-Digest
 * to); that is not an error. A request already carrying a Signature is CHAINED
 * onto (appendSignature), never replaced. Naming stays family-consistent with the
 * newXResolver siblings; the whole-surface newX -> create rename is a separate,
 * broader change, not this module's concern.
 */
export function createSigningTransport<R>(
	send: OutboundSend<R>,
	opts: SigningTransportOptions,
): OutboundSend<R> {
	const window = opts.window ?? defaultWindow();
	return async (url, init) => {
		const body = init.body;
		const method = init.method ?? "GET";
		const headers = init.headers ?? {};

		// No body / predicate-excluded: pass through UNSIGNED, body untouched.
		const excluded =
			opts.predicate !== undefined &&
			!opts.predicate({ url, method, headers, body });
		if (body === undefined || excluded) {
			return send(url, init);
		}

		// Signature-Agent SET-IF-ABSENT: the incoming header (an upstream sig1's
		// covered directory) wins; otherwise the transport stamps its own.
		const existingAgent = getHeader(headers, SignatureAgentHeader);
		const signatureAgent = existingAgent ?? opts.signatureAgent ?? "";

		const prior: PriorSignatures = {
			signatureInput: getHeader(headers, "signature-input") ?? "",
			signature: getHeader(headers, "signature") ?? "",
		};

		const signed = await signOutbound({
			privKey: opts.privKey,
			keyid: opts.keyid,
			method,
			url,
			body,
			authorization: getHeader(headers, "authorization") ?? "",
			signatureAgent,
			window,
			appendOnly: opts.appendOnly ?? false,
			prior,
		});

		// Forward the SAME body bytes (body integrity — buffer for the digest
		// but never consume/replace the payload; Go resets req.Body + GetBody).
		return send(url, {
			...init,
			headers: mergeSigned(headers, signed.headers),
			body,
		});
	};
}

File: sdk/python/ramp_sdk/signing_transport.py
"""sdk/python outbound-sign face — TRANSPORT-NEUTRAL, the client analogue of the
Go RoundTripper sign face.

``SigningTransport.sign_outbound`` assembles the RFC 9421 headers for one outbound
request (RFC 9530 Content-Digest over the EXACT body bytes) via the shipped L1
``httpsig.sign_request`` byte oracle; it adds NO new crypto and performs NO IO. It
is NOT an HTTP client binding: it imports no httpx and wraps no transport — the
consumer (e.g. the app MCP shim, src/mcp) owns the actual httpx transport and
applies the returned headers itself. A returned offer is verified through the core
Verifier (in the app, not here).
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import TYPE_CHECKING

from ramp_sdk.core import sign_offer_acceptance_jcs
from ramp_sdk.httpsig import sign_request
from ramp_sdk.pop import sign_agent_binding
from ramp_sdk.window import Window, clock_window

if TYPE_CHECKING:
    from collections.abc import Callable

# Default proof window (seconds) for the created/expires params when the caller does
# not inject an explicit window.
_DEFAULT_TTL_SEC = 600


@dataclass(frozen=True)
class SignedOutbound:
    """The signed outbound request: the RFC 9421 headers to attach to the httpx send."""

    method: str
    url: str
    body: bytes
    headers: dict[str, str]


class SigningTransport:
    """Signs outbound requests via the core sign seam (RFC 9421 over exact body bytes).

    The signer seed + keyid are injected; the clock is injectable for determinism.
    ``sign_outbound`` returns the request plus EVERY covered header at the value that
    entered the signature base — Content-Digest / Signature-Input / Signature, and
    Authorization / Signature-Agent whose values may be empty. The seam MCP wires into
    its httpx transport.
    """

    def __init__(
        self,
        *,
        signer_seed: bytes,
        keyid: str,
        now: Callable[[], float] | None = None,
        ttl_sec: int = _DEFAULT_TTL_SEC,
        signature_agent: str = "",
        window: Window | None = None,
    ) -> None:
        """``signature_agent`` is the signer's own WBA directory URL.

        Signature-Agent is a COVERED wire component: the sign seam always binds
        its value (empty included) into the signature base, and the header is
        always attached carrying that same value — empty included. Binding it
        without sending it is not binding it, because the verifier rebuilds the
        base from the request it received and refuses one whose covered name has
        nothing on the wire under it. When non-empty it is also how the verifier
        resolves the signer's keys against that directory.

        Always-stamp is the client shape — Python has no relay path, so the Go
        relay's set-if-absent guard (core.WithSignatureAgent) has no analogue
        here.
        """
        self._signer_seed = signer_seed
        self._keyid = keyid
        self._now = now or time.time
        self._ttl_sec = ttl_sec
        self._signature_agent = signature_agent
        # Source (created, expires) from an injected Window, defaulting to a
        # clock_window over now/ttl_sec. clock_window int()-truncates to whole
        # seconds, so the @signature-params bytes stay byte-identical to the
        # historical inline ``int(self._now())`` mint.
        self._window = window or clock_window(self._now, self._ttl_sec)

    def sign_offer_acceptance(
        self,
        *,
        offer_sig: str,
        requester_id: str,
        requester_domain: str,
        idempotency_key: str,
    ) -> tuple[str, str]:
        """Sign the detached acceptance a purchase carries; return ``(hex_signature, alg)``.

        It signs with the REQUEST signer's key, and there is deliberately no option for a
        separate acceptance key. The protocol carries one agent identity:
        ``agent_identity_hash`` is defined as the thumbprint of the agent's request-signing
        key, an Exchange verifies the detached acceptance against the key registered for
        the caller its request signature identified, and the delivery URL is bound to that
        same thumbprint. A second key would be refused at execute, and any URL it did
        produce could never be fetched — the presented key would not match the binding.

        It lives on the transport so the key stays in one place: the client composes this
        seam and never handles key material, which is the shape the Go oracle has, where
        the acceptance is signed by passing the injected Signer itself.
        """
        return sign_offer_acceptance_jcs(
            seed=self._signer_seed,
            offer_sig=offer_sig,
            requester_id=requester_id,
            requester_domain=requester_domain,
            idempotency_key=idempotency_key,
        )

    def sign_agent_binding(self, *, url: str, window: Window) -> tuple[str, str, str]:
        """Mint the proof of possession for one bound delivery GET; return the agent-key
        header value, the Signature-Input and the Signature.

        The SAME key as the request signer, for the same reason as
        :meth:`sign_offer_acceptance`: the delivery URL is bound to the thumbprint of the
        agent's request-signing key, so a proof minted under any other key presents an
        identity the URL was not issued to.
        """
        created, expires = window()
        return sign_agent_binding(
            url=url, signer_seed=self._signer_seed, created=created, expires=expires
        )

    def sign_outbound(
        self,
        *,
        method: str,
        url: str,
        body: bytes,
        authorization: str,
    ) -> SignedOutbound:
        """Sign an outbound request; return it with the RFC 9421 headers attached.

        Authorization is always bound — pass an empty string to pin its absence
        (mirror the L1 sign_request contract).
        """
        created, expires = self._window()
        signed = sign_request(
            method=method,
            url=url,
            body=body,
            authorization=authorization,
            signer_seed=self._signer_seed,
            keyid=self._keyid,
            created=created,
            expires=expires,
            signature_agent=self._signature_agent,
        )
        # EVERY covered header, at exactly the value that entered the signature base —
        # empty values included. See docs/design-history.md,
        # "A covered header the peer never receives is not bound", for why binding one
        # without sending it is not binding it.
        #
        # Taken straight off ``signed``, never re-read from the arguments: the primitive
        # echoes what it bound, so there is one place the emitted value can come from and
        # no way for the two to drift.
        return SignedOutbound(
            method=method,
            url=url,
            body=body,
            headers={
                "content-digest": signed.content_digest,
                "signature-input": signed.signature_input,
                "signature": signed.signature,
                "authorization": signed.authorization,
                "signature-agent": signed.signature_agent,
            },
        )
